network/framework/src/application/storage.rs

```diff
@@ -16,99 +16,145 @@ use aptos_config::{
+            subscribers: RwLock::new(vec![]),
         };
 
         // Initialize each network mapping and trusted peer set
-        network_ids.iter().for_each(|network_id| {
-            // Update the peers and metadata map
-            peers_and_metadata
-                .peers_and_metadata
-                .write()
-                .insert(*network_id, HashMap::new());
-
-            // Update the trusted peer set
-            peers_and_metadata.trusted_peers.insert(
-                *network_id,
-                Arc::new(ArcSwap::from(Arc::new(PeerSet::new()))),
-            );
-        });
-
-        // Initialize the cached peers and metadata
-        let cached_peers_and_metadata = peers_and_metadata.peers_and_metadata.read().clone();
-        peers_and_metadata.set_cached_peers_and_metadata(cached_peers_and_metadata);
-
-        // Return the peers and metadata container
+        {
+            let mut wat = peers_and_metadata.peers_and_metadata.write();
+            peers_and_metadata.network_ids.iter().for_each(|network_id| {
+                wat.insert(*network_id, HashMap::new());
+
+                peers_and_metadata
+                    .trusted_peers
+                    .insert(*network_id, Arc::new(RwLock::new(PeerSet::new())));
+            });
+        }
+
         Arc::new(peers_and_metadata)
     }
 
+    /// Returns the networks currently held in the container
+    pub fn get_registered_networks(&self) -> impl Iterator<Item = NetworkId> + '_ {
+        self.network_ids.iter().copied()
+    }
+
     /// Returns all peers. Note: this will return disconnected and unhealthy peers, so
     /// it is not recommended for applications to use this interface. Instead,
     /// `get_connected_peers_and_metadata()` should be used.
-    pub fn get_all_peers(&self) -> Result<Vec<PeerNetworkId>, Error> {
-        // Get the cached peers and metadata
-        let cached_peers_and_metadata = self.cached_peers_and_metadata.load();
-
-        // Collect all peers
+    pub fn get_all_peers(&self) -> Vec<PeerNetworkId> {
         let mut all_peers = Vec::new();
-        for (network_id, peers_and_metadata) in cached_peers_and_metadata.iter() {
-            for (peer_id, _) in peers_and_metadata.iter() {
+        let read = self.peers_and_metadata.read();
+        for (network_id, peers) in read.iter() {
+            for (peer_id, _) in peers.iter() {
                 let peer_network_id = PeerNetworkId::new(*network_id, *peer_id);
                 all_peers.push(peer_network_id);
             }
         }
-        Ok(all_peers)
+        all_peers
     }
 
-    /// Returns metadata for all peers currently connected to the node
-    pub fn get_connected_peers_and_metadata(
+    /// Return copy of list of peers if generation is different than previously held.
+    /// May optionally filter for connected peers only.
+    /// May optionally filter on match for any of a set of ProtocolId.
+    pub fn get_all_peers_generational(
```

Comment:
I imagine the generational changes could be separated out?

Reply:
Yup. Did that, but then replaced it all with the ArcSwap stuff from `main`